### PR TITLE
[#1802] improvement(netty): Allow sharing Netty's memory allocators

### PR DESCRIPTION
**Pull request by rickyma**
### What changes were proposed in this pull request?

Introducing shared ByteBuf allocators.
This feature can be enabled via the `rss.client.netty.client.shared.allocator.enabled` configuration.

If enabled then only three ByteBuf allocators are created: 
one PooledByteBufAllocator where caching is allowed, one PooledByteBufAllocator where not and one UnpooledByteBufAllocator When disabled, a new allocator is created for each transport client.

We set the default value to true to reduce memory usage, refer to https://github.com/apache/spark/pull/23278.

### Why are the changes needed?

For: https://github.com/apache/incubator-uniffle/issues/1802.

### Does this PR introduce _any_ user-facing change?

The `rss.client.netty.client.shared.allocator.enabled` configuration is introduced.

### How was this patch tested?

Existing UTs.

### CODE REVIEW

File: common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java [hunk 1]
```diff
@@ -117,14 +172,6 @@ public static String getServerConnectionInfo(Channel channel) {
     return String.format("[%s -> %s]", channel.localAddress(), channel.remoteAddress());
   }
 
-  private static class AllocatorHolder {
```
Comment (rickyma) [inline common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java:120]: Remove this to unify the code.

**Comment by rickyma** [comment]
@jerqi @zuston PTAL.

**Review by jerqi — APPROVED** [review]
LGTM.

File: common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java [hunk 2]
```diff
@@ -43,6 +43,13 @@ public class NettyUtils {
 
   private static final long MAX_DIRECT_MEMORY_IN_BYTES = PlatformDependent.maxDirectMemory();
 
+  /** Specifies an upper bound on the number of Netty threads that Uniffle requires by default. */
+  private static int MAX_DEFAULT_NETTY_THREADS = 8;
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java:47]: Why 8 as the upper bound?

Comment (rickyma) [reply common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java:47]: The codes are copied from Spark, useless comments are removed. Although it says upper bound, but actually it is a way to ensure **_minimum_** concurrency. 
Refer to https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java#L50.

Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/util/NettyUtils.java:47]: Got it.

**Review by zuston — APPROVED** [review]
lgtm

**Pull request merged**